src/plugins/dist_matrixops/dist_diag.cpp

```diff
@@ -0,0 +1,576 @@
+                    if (des_start < cur_stop && des_stop > cur_start)
+                    {
+                        auto indices = detail::diag_index_calculation_1d(
+                            des_start, des_stop, cur_start, cur_stop);
+                        blaze::subvector(res_arr, indices.projected_start_,
+                            indices.intersection_size_) = blaze::subvector(v,
+                                indices.local_start_,
+                                indices.intersection_size_);
+                    }
+
+                    // collecting other parts from remote localities
+                    for (std::uint32_t loc = 0; loc != num_localities; ++loc)
+                    {
+                        if (loc == loc_id)
+                        {
+                            continue;
+                        }
+
+                        // the array span in for loc-th locality
+                        tiling_span loc_span =
+                            arr_localities.tiles_[loc].spans_[span_index];
+
+                        auto indices = detail::diag_retile_calculation_1d(
+                            loc_span, des_start, des_stop);
+                        if (indices.intersection_size_ > 0)
+                        {
+                            // loc_span has the part of result that we need
+                            blaze::subvector(res_arr, indices.projected_start_,
+                                indices.intersection_size_) =
+                                v_data
+                                .fetch(loc, indices.local_start_,
+                                    indices.local_start_ +
+                                    indices.intersection_size_)
+                                .get();
+                        }
+                    }
+
+                    blaze::band(result, num_band) = res_arr;
+                }
+
+                else // the new array is a subset of the original array
+                {
+                    blaze::band(result, num_band) =
+                        blaze::subvector(v, rel_start, des_size);
+                }
+            }
+        }
+        else
+        {
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "dist_diag::dist_diag1d",
+                generate_error_message(
+                    "wrong numtiles input when tiling_type is column"));
+        }
+
+        tiling_information_2d tile_info(
+            tiling_span(row_start, row_start + row_size),
+            tiling_span(column_start, column_start + column_size));
+
+        auto locality_ann = arr_localities.locality_.as_annotation();
+        auto attached_annotation =
+            std::make_shared<annotation>(localities_annotation(locality_ann,
+                tile_info.as_annotation(name_, codename_),
+                arr_localities.annotation_, name_, codename_));
+
+        return primitive_argument_type(result, attached_annotation);
+    }
```

Comment:
Could you extract the various cases into separate functions, please? This function is very large and difficult to read.

Reply:
Will do it.